### PR TITLE
10bit rdoq

### PR DESCRIPTION
**Pull request by joelsole**
## Description

Add RDOQ on the 10-bit path
* Added config parameter 

Associated to issue #622

## Type of change

Added feature

## Tests and performance

BD-rate gains for  30 8-bit sequences and twelve 10-bit test sequences is:

Bitdepth | PSNRY | PSNRU | PSNRV | CPSNRYUV | TruePSNRY | TruePSNRU | TruePSNRV | TruePSNRYUV | BDVMAF | EncTimeInc
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
8 | 0 | 0.001 | 0 | 0 | 0 | 0.001 | 0 | 0 | 0 | -0.286
10 | -1.097 | 2.187 | 2.169 | -0.718 | -1.11 | 2.223 | 2.206 | -1.104 | -2.362 | 4.178

4% encoding time increase for 10-bit. No change in memory.
No changes for 8-bit.

### CODE REVIEW

**Comment by joelsole** [comment]
The config / cli in this PR (specifcially, AUTO_MODE) follows the guidelines in the branch https://github.com/OpenVisualCloud/SVT-AV1/commits/tools_on_off

**Comment by grill05** [comment]
@joelsole This is not directly related to this change, but could you please give a sample command you are using for testing BD-rate change/speed change for 10-bit(in PR's referenced by this one)? And are the tests with a 10-bit yuv input or via pipes?
I ask because I keep getting crashes when I try to verify the speed changes for 10-bit :)

**Comment by okhlif** [comment]
@joelsole  Are you seeing loss in PSNR U and V ? 
I ran RDOQ ON/OFF in context of Fullset objective fast 1 in 10bit with **hbd 0** and I am seeing gains both in luma and chroma as follow 

Y | U | V
-- | -- | --
-1.3% | -1.7% | -1.5%

**Comment by joelsole** [comment]
> @joelsole Are you seeing loss in PSNR U and V ?
> 
> EDITED
> 
> Actually it is the reverse, hbd 0 RDOQ seems bringing loss.

Yes, that makes more sense. I was scratching my head to your initial post.

To be sure we are on the same page:
 - 10b content with 10b RDOQ ON for MD and for encdec brings gains
 - 8b  content with 8b RDOQ ON for  MD and for encdec brings gains
 - 10b content with 8b RDOQ ON for MD and 10b RDOQ ON for encdec has losses

Is this what you are observing? If so, were you expecting seeing gains by adding RDOQ for the 8b MD but doing the final RDOQ for 10b?

**Comment by joelsole** [comment]
@okhlif any further comment on the above?

If the concern is the interaction between the 8bit-RDOQ in the mode decision pass when hbd is 0 and the 10bit-RDOQ in the encoding pass, we could add a safeguard like follows:

 `if (sequence_control_set_ptr->static_config.encoder_bit_depth > 8
      && picture_control_set_ptr->hbd_mode_decision==0)
         perform_rdoq = EB_FALSE;`

Otherwise, is this good to go?

File: Source/Lib/Common/Codec/aom_dsp_rtcd.c [hunk 1]
```diff
@@ -412,6 +412,9 @@ void setup_rtcd_internal(EbAsm asm_type)
     eb_av1_quantize_fp_64x64 = eb_av1_quantize_fp_64x64_c;
     if (flags & HAS_AVX2) eb_av1_quantize_fp_64x64 = eb_av1_quantize_fp_64x64_avx2;
 
+    eb_av1_highbd_quantize_fp = eb_av1_highbd_quantize_fp_c;
+    //if (flags & HAS_AVX2) eb_av1_highbd_quantize_fp = eb_av1_highbd_quantize_fp_avx2;
```
Comment (hassount) [inline Source/Lib/Common/Codec/aom_dsp_rtcd.c]: so no avx2 ?

**Comment by okhlif** [comment]
> > @joelsole Are you seeing loss in PSNR U and V ?
> > EDITED
> > Actually it is the reverse, hbd 0 RDOQ seems bringing loss.
> 
> Yes, that makes more sense. I was scratching my head to your initial post.
> 
> To be sure we are on the same page:
> 
> * 10b content with 10b RDOQ ON for MD and for encdec brings gains
> * 8b  content with 8b RDOQ ON for  MD and for encdec brings gains
> * 10b content with 8b RDOQ ON for MD and 10b RDOQ ON for encdec has losses
> 
> Is this what you are observing? If so, were you expecting seeing gains by adding RDOQ for the 8b MD but doing the final RDOQ for 10b?

Sorry Joel for the delay yes that what I am seeing for 10bit input ( I did not test 8 bit input) 
We never saw a feature in 8bit MD that causes loss for 10bit. 
Did you actually test lately 8b  content with 8b RDOQ ON for  MD ?

File: Source/Lib/Common/Codec/aom_dsp_rtcd.c (hunk 1 above)
Comment (joelsole) [reply Source/Lib/Common/Codec/aom_dsp_rtcd.c]: yes, there is avx2. This was commented out for testing. Removed.

**Comment by joelsole** [comment]
> @okhlif any further comment on the above?
> 
> If the concern is the interaction between the 8bit-RDOQ in the mode decision pass when hbd is 0 and the 10bit-RDOQ in the encoding pass, we could add a safeguard like follows:
> 
> `if (sequence_control_set_ptr->static_config.encoder_bit_depth > 8 && picture_control_set_ptr->hbd_mode_decision==0) perform_rdoq = EB_FALSE;`
> 
> Otherwise, is this good to go?

**Comment by joelsole** [comment]
> Did you actually test lately 8b content with 8b RDOQ ON for MD ?

Yes, I did test 8b content with 8b RDOQ ON in the MD: it is bringing the expected gains. 

Note that the losses appear (for 8b) when RDOQ is OFF for all MD stages but ON for encdec.

In the upcoming commit, I will disable the undesired combination of RDOQ ON for 8b MD with hbd 0.

**Comment by okhlif** [comment]
> > Did you actually test lately 8b content with 8b RDOQ ON for MD ?
> 
> Yes, I did test 8b content with 8b RDOQ ON in the MD: it is bringing the expected gains.
> 
> Note that the losses appear (for 8b) when RDOQ is OFF for all MD stages but ON for encdec.
> 
> In the upcoming commit, I will disable the undesired combination of RDOQ ON for 8b MD with hbd 0.

Thank you.  RDOQ ON loss for 8b MD with hbd  0 might be addressed with an upcoming fix to ATB.

File: Source/App/EncApp/EbAppConfig.c [hunk 2]
```diff
@@ -959,6 +965,12 @@ static EbErrorType VerifySettings(EbConfig *config, uint32_t channelNumber)
         fprintf(config->error_log_file, "Error instance %u: Invalid OBMC flag [0 - 1], your input: %d\n", channelNumber + 1, config->target_socket);
         return_error = EB_ErrorBadParameter;
     }
+
+    // RDOQ
+    if (config->enable_rdoq != 0 && config->enable_rdoq != 1 && config->enable_rdoq != -1) {
```
[the remark this answers is not in the document]
Comment (joelsole) [reply Source/App/EncApp/EbAppConfig.c]: adressed

**Pull request merged**